### DISCUSSION TITLE
Firmata-like protocol for WebUSB on TinyUSB?

### DISCUSSION THREAD

**Discussion by ATMakersBill** [hathach/tinyusb#580] [Ideas]
Hi folks,

First thanks for TinyUSB - it's so much improved over the past year!

I am aware of WebSerial, however, because it's not a standard feature, as well as not wanting to go through the overhead of JSON or the like I am looking for a true WebUSB protocol that we can use to read from sensors and configure microcontroller-based systems.

Ideally, this protocol would allow the browser (Javascript) to:

1. Read or set any digital or analog pin
2. Read or set other "meta" values that the device wants to expose (i.e. CPX's accelerometer data)
3. Query the capabilities of the device
4. Request that the device start and stop capturing high resolution values (given sample rate, pin #)
5. Request that the device send the results of the last capture
6. Request that the device start notifying on pin values (i.e. the browser would go into "listen mode" and get asynchronous notifications when enabled pins go high or low.

This feels a bit like Firmata, so I thought there probably is some groundwork laid here - perhaps I shouldn't start from scratch :-)

Does anyone know of something like this for WebUSB?  Can Firmata run over WebUSB?  Does anyone want to play with me on this if we do start from scratch?  It seems like something that would be very handy for things like the Circuit Playground Express.

You're welcome to tell me this isn't a TinyUSB idea, but it's one of the easiest stacks to use on WebUSB and this seems like the kind of protocol that would be easy to expose up to the CircuitPython layer, etc.

Bill

**Comment by ladyada** [comment]
this is a massive amount of work to take on. if you'd like, you can try! but as webserial is just as non-standard as webusb, we recommend just use webserial with firmata.

**Reply by ATMakersBill** [reply]
Isn't webserial still in an origin trial?  I realize that only Chrome supports WebUSB, but I've been able to connect to the CPX (and CPB) without enabling any experimental features.  That will be important for what we're trying to do.

**Comment by chegewara** [comment]
Each protocol has its own pros and cons (webserial and webusb) and both are still experimental and available on chrome only. Personally i prefer webusb, because of issues with webserial on linux, but i have to agree, it is big project to do it with webusb or webserial. It would be 2 parts project:
- hosted website with full GUI to control device,
- device interface that will work with this website, you can create firmata like API.

**Reply by chegewara** [reply]
Here s small example that is reading DS18D20 on esp32S2 and displaying on webusb:
https://esp32.com/viewtopic.php?f=18&t=15908&p=63005#p61932

**Reply by ATMakersBill** [reply]
Interesting how similar our interfaces are :-) 
![image](https://user-images.githubusercontent.com/32851154/103470920-08c3d280-4d47-11eb-90b7-99e0a176556c.png)
That's a sample reading the X coordinates from a CPX and plotting them with Chart.JS

**Reply by ATMakersBill** [reply]
It just felt like putting a higher layer above WebUSB might make some sense since we're all looking to do the same thing.

**Reply by chegewara** [reply]
Yes. Im not good at html/js, so i found chartJS example and basically copy/paste it and i am feeding data from webusb. Its a simple example to shows me how it works.
Webserial is as good as webusb in that case, but linux is claiming device with cdc_acm module.

**Comment by hathach** [comment]
It is a great idea. But it is at application level and not part of the usb stack. WebUSB to tinyusb is a stream data, so yeah I think you could use any (existing) open protocol on top of it

**Comment by ATMakersBill** [comment]
For those of you wondering why I'm investigating this (and why it's important) here's what we're considering right now on the IMPAct project (an initiative ATMakers is running in 2021 with several universities)
https://drive.google.com/file/d/1j2LLjm_WxZxXysQ80rzZmc9foYEoohB7/view?usp=sharing
